tensorflow/compiler/xla/service/cpu/onednn_rewriter.cc

```diff
@@ -0,0 +1,125 @@
+/* Copyright 2023 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+#if defined(INTEL_MKL) && defined(ENABLE_ONEDNN_V3)
+
+#include "tensorflow/compiler/xla/service/cpu/onednn_rewriter.h"
+
+#include "tensorflow/compiler/xla/hlo/ir/dfs_hlo_visitor_with_default.h"
+#include "tensorflow/compiler/xla/hlo/ir/hlo_instruction.h"
+#include "tensorflow/compiler/xla/service/cpu/backend_config.pb.h"
+#include "tensorflow/compiler/xla/service/cpu/onednn_memory_util.h"
+#include "tensorflow/compiler/xla/service/pattern_matcher.h"
+#include "tensorflow/compiler/xla/status_macros.h"
+
+namespace xla {
+namespace cpu {
+
+namespace {
+namespace m = match;
+
+Status ValidateDotDimensionNumbers(const DotDimensionNumbers& dim_numbers) {
+  // Checks some invariants that do not hold in general, but DotDecomposer
+  // should have established for us.
+  TF_RET_CHECK(dim_numbers.lhs_contracting_dimensions_size() == 1);
+  std::vector<int64_t> batch_dim_numbers(
+      dim_numbers.lhs_batch_dimensions_size());
+  absl::c_iota(batch_dim_numbers, 0);
+  TF_RET_CHECK(
+      absl::c_equal(batch_dim_numbers, dim_numbers.lhs_batch_dimensions()));
+  TF_RET_CHECK(
+      absl::c_equal(batch_dim_numbers, dim_numbers.rhs_batch_dimensions()));
+  return OkStatus();
+}
+
+}  // namespace
+
+class OneDnnRewriterVisitor : public DfsHloRewriteVisitor {
+ public:
+  // Matches patterns for possible MatMul fusions that are supported by oneDNN
+  // library. Matched hlo instruction(s) are replaced by custom call.
+  Status HandleDot(HloInstruction* instr) override {
+    // Currently, blocking control dependencies
+    if (instr->HasControlDependencies()) return OkStatus();
+    HloInstruction* dot_instr;
+    auto pattern = m::Op(&dot_instr).WithOpcode(HloOpcode::kDot);
+    if (!Match(instr, pattern)) return OkStatus();
+
+    // The rewrite pass runs after dot-decomposition pass. Adjust
+    // the rewrite condition when the rewrite pass is moved before
+    // dot-decomposition pass.
+
+    // Currently, we rewrite when the data type is F32 or BF16. Note we do not
+    // need to check equality of contraction dim-size of the operands. HLO
+    // verifier already does the job. We however, need to check if contraction
+    // is over only 1 dimension (a.k.a. K dimension in matrix-multiplication
+    // parlance). We also restrict batch dimensions of the operands mathces.
```

Comment:
Nit: Grammar and typo.
```suggestion
    // verifier already does the job. We, however, need to check if contraction
    // is over only 1 dimension (a.k.a. K dimension in matrix-multiplication
    // parlance). We also restrict batch dimensions of the operands matches.
```